sql/catalyst/src/main/scala/org/apache/spark/sql/types/numerics.scala

```diff
@@ -121,8 +121,8 @@ object FloatExactNumeric extends FloatIsFractional {
   private def overflowException(x: Float, dataType: String) =
     throw new ArithmeticException(s"Casting $x to $dataType causes overflow")
 
-  private val intUpperBound = Int.MaxValue.toFloat
-  private val intLowerBound = Int.MinValue.toFloat
+  private val intUpperBound = Int.MaxValue
```

Comment:
Hm, I'm also not clear how this helps - won't it just promote to a float in the comparison below anyway?
Do we want floorDiv, etc, instead?

Reply:
`Math.floor` returns double, so it's promoted to double

Reply:
As mentioned by cloud-fan, it seems that an cast int to float, then casted to double is not same with casting it to double directly.
![image](https://user-images.githubusercontent.com/6757692/72084214-0e5a2880-333e-11ea-820d-dc4172d93986.png)

Reply:
it's true
```
scala> Int.MaxValue.toDouble
res2: Double = 2.147483647E9

scala> Int.MaxValue.toFloat.toDouble
res3: Double = 2.147483648E9
```